core/jvm/src/main/scala/zio/internal/OneShot.scala

```diff
@@ -45,18 +45,39 @@ private[zio] class OneShot[A] private (@volatile var value: A) {
 
   /**
    * Retrieves the value of the variable, blocking if necessary.
+   *
+   * @param timeout The maximum amount of time the thread will be blocked.
+   * @throws Error if the timeout is reached without the value being set.
    */
-  final def get(timeout: Long = Long.MaxValue): A = {
-    if (value == null) {
+  final def get(timeout: Long): A = {
+    var remaining = timeout
+    while (value == null && remaining > 0L) {
+      val start = System.currentTimeMillis()
       this.synchronized {
-        if (value == null) this.wait(timeout)
+        if (value == null) this.wait(remaining)
       }
-
-      if (value == null) throw new Error("Timed out waiting for variable to be set")
+      remaining -= System.currentTimeMillis() - start
```

Comment:
nano

Reply:
❓ I don't follow.

Reply:
@hepin1989 We shouldn't use System.nanotime as it can provide negative values: https://stackoverflow.com/questions/510462/is-system-nanotime-completely-useless/54566928#54566928

Reply:
Now that I look into it, we _should_ use `nanoTime`. Yes, it can return negative values, but that doesn't matter because all it can be used for is subtraction from another `nanoTime` reading. The difference is all we need to know.

I hadn't previously realised that `currentTimeMillis` is not ideal for measuring elapsed time, as it is a wall-clock time that can shift around if the system time is corrected.

Unfortunately the `wait` API doesn't have a variant taking a single nanoseconds long parameter 😠 .